### DISCUSSION TITLE
[Flow Manager] Behavior in version 2.5.0

### DISCUSSION THREAD

**Discussion by cliviero** [mia-platform/community#540] [Q&A] [answered]
### What product are you having troubles with?

Marketplace

### What Console version are you using?

1.104.0

### Description

Hi everyone,

I'm experiencing what I believe is a concurrency issue with the Flow Manager in version 2.5.0, and I would appreciate your help in understanding its internal behavior better.

In my project, the Flow Manager acts as both a consumer and a producer and is integrated with a client that interacts with it. However, I've been repeatedly seeing the following error in the logs in a somewhat random manner:

`Unable to process event, since it was unexpected in current state`

This error occurs intermittently; sometimes, the Flow Manager changes states correctly, and other times it doesn't. Upon reviewing the log history, I noticed that there are only a few milliseconds between state changes. This leads me to believe that the issue might be related to how quickly the state is being saved to MongoDB, which could be affecting the processing of subsequent commands.

My question is: Does the Flow Manager first save its state and then attempt to execute the new command, or does it send both requests (saving the state and executing the command) in parallel? In other words, could it be that the Flow Manager is attempting to change state and execute a new command simultaneously without waiting for the new state to be fully saved?

Any explanation or suggestions on how to handle this situation would be greatly appreciated.

Thanks in advance!

**Comment by frengooo** [comment, marked as answer]
Good morning @cliviero,
thanks for reaching out.

The Flow Manager sends the commands only after the transition to the new state has been completed and saved on the database.
The flow is the following one:
* the Flow Manager receives the event related to a saga;
* the Flow Manager gets the saga from the database;
* the Flow Manager verifies if the received event is expected, in the current state of the saga (based on how you define the Flow Manager saga);
* if the event is expected, the Flow Manager updates locally the saga JSON;
* the Flow Manager saves the new saga to the database;
* the Flow Manager sends the commands, if any, related to the new state.

The `Unable to process event, since it was unexpected in current state` error is caused by an unexpected event received on a saga.
Each saga state supports only the events that are expected in the specific state the saga is in a specific moment; if the Flow Manager receives an event that is not defined in that state, it is unable to handle it and to update the saga state, for this reason:
* the Flow Manager logs the error;
* the Flow Manager leaves the saga unchanged.

Tell me if this explanation is enough or if you need other support 😄
